### PR TITLE
[Prototype] Bifrost integration

### PR DESCRIPTION
**Pull request by DnsZhou**

### CODE REVIEW

File: src/hooks/useBifrostWebSocketEffect.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+/**
+ * WebSocket message listener hook.
+ *
+ * @param callback - Callback method, WebSocket parsed data attribute will be passed in
+ * @param options - See {@link IUseBifrostWebSocketEffectOptions}
+ */
+export const useBifrostWebSocketEffect = (
+  callback: (wsData: any) => void,
```
Comment (matedo1) [inline src/hooks/useBifrostWebSocketEffect.ts:52]: Argument name could match the one which is used later, so in this case `message: string` or more specific `logLine: string` in this case.

File: src/components/BuildLogPage/BuildLogPage.tsx [hunk 2]
```diff
@@ -23,12 +26,24 @@ export const BuildLogPage = () => {
...
+
   useEffect(() => {
     if (!isBuilding) {
       serviceContainerBuildLogRunner({ serviceData: { id: buildId } });
     }
   }, [serviceContainerBuildLogRunner, buildId, isBuilding]);
 
+  useBifrostWebSocketEffect(
+    useCallback(
+      (message: string) => {
```
Comment (matedo1) [inline src/components/BuildLogPage/BuildLogPage.tsx:39]: `logLine: string` could be considered

File: src/components/BuildLogPage/BuildLogPage.tsx [hunk 3]
```diff
@@ -23,12 +26,24 @@ export const BuildLogPage = () => {
 
   const logData = useMemo(() => serviceContainerBuildLog.data?.split(/[\r\n]/) || [], [serviceContainerBuildLog.data]);
 
+  const [buffer, addLines] = useDataBuffer(0, timestampHiglighter);
```
Comment (matedo1) [inline src/components/BuildLogPage/BuildLogPage.tsx:29]: More specific naming like `[logBuffer, addLogLines]` should be considered.

**Review by matedo1 — CHANGES_REQUESTED** [review]
Hi @DnsZhou , here I am providing some feedback, tests are not passing at this moment, can you please fix them?

File: src/hooks/useBifrostWebSocketEffect.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+interface IUseBifrostWebSocketEffectOptions {
+  buildId?: string;
+
+  // condition based listening, when true, events will be ignored (for example when Build is finished)
+  preventListening?: boolean;
+
+  // debug identification string
+  debug?: string;
+}
```
Comment (patrikk0123) [inline src/hooks/useBifrostWebSocketEffect.ts:43]: I would try to unite the common fields with `IUsePncWebSocketEffectOptions`

File: src/components/BuildLogPage/BuildLogPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/BuildLogPage/BuildLogPage.tsx:29]: I would add a delay to the buffer, 300ms or more. I reviewed the whole PR and seems to be fine.

[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/BuildLogPage/BuildLogPage.tsx:29]: I would add a delay to the buffer, 300ms or more. I reviewed the whole PR and seems to be fine.

**Comment by DnsZhou** [comment]
Replaced by https://github.com/project-ncl/pnc-web-ui-react/pull/575

**Pull request closed without merging**